network/eventManager.go

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2017 The Alvalor Authors
+//
+// This file is part of Alvalor.
+//
+// Alvalor is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Alvalor is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with Alvalor.  If not, see <http://www.gnu.org/licenses/>.
+
+package network
+
+import (
+	"time"
+
+	"github.com/rs/zerolog/log"
+)
+
+type eventManager interface {
+	Disconnected(addr string)
+	Connected(addr string)
+	Received(addr string, msg interface{})
+}
+
+type simpleEventManager struct {
+	subscriber chan interface{}
+}
+
+func (mgr *simpleEventManager) Disconnected(address string) {
+	select {
+	case mgr.subscriber <- Disconnected{Address: address, Timestamp: time.Now()}:
+	default:
+		log.Debug().Msg("subscriber stalling")
+	}
+}
+
+func (mgr *simpleEventManager) Connected(address string) {
+	select {
+	case mgr.subscriber <- Connected{Address: address, Timestamp: time.Now()}:
+	default:
+		log.Debug().Msg("subscriber stalling")
+	}
+}
+
+func (mgr *simpleEventManager) Received(address string, msg interface{}) {
+	select {
+	case mgr.subscriber <- Received{Address: address, Message: msg, Timestamp: time.Now()}:
+	default:
+		log.Debug().Msg("subscriber stalling")
+	}
```

Comment:
I think it would be cleaner to create a single function `Event(event interface{})`, and to call that from the others, so you only have the select in one place instead of three. Maybe you even want to do the signature with an `error` return and you can return subscriber stalling as an error and bubble it up all the way to the handler, where you log it.

Reply:
Agree about the ability to push error to handler. Regarding single Event, currently, all functions set Timestamp property so it makes Connected/Received/Disconnected more simple.

Reply:
That is fine, but you can still do something like this:
```
event := Connected{Address: address, Timestamp: time.Now()}
return mgr.Event(event)
```
(considering there are error returns)

The point is simply to have a single select statement, rather than duplicate that.

Reply:
Removed duplicates.